gym_electric_motor/core.py

```diff
@@ -755,4 +685,83 @@ def on_step_end(self):
+
+    **Soft Constraints:**
+        To enable a higher flexibility, the constraints return a violation degree (float) instead of a simple violation
+        flag (bool). So, even before the limits are violated, the reward function can take the limit violation degree
+        into account. If the violation degree is at 0.0, no states are in a dangerous region. For values between 0.0 and
+        1.0 the reward will be decreased gradually so that the agent will learn to avoid these state regions.
+        If the violation degree reaches 1.0 the episode is terminated.
+
+    **Hard Constraints:**
+        With the above concept, also hard constraints that directly terminate an episode without any "danger"-region
+        can be modeled. Then, the violation degree of the constraint directly changes from 0.0 to 1.0, if a violation
+        occurs.
+
+    """
+
+    def __init__(self,  limit_constraints=(), additional_constraints=(), merge_violations='max'):
+        """
+        Args:
+            limit_constraints(list(str)/'all'):
+                Shortcut parameter to pass all states that limits shall be observed.
+                    - list(str): Pass a list with state_names and all of the states will be observed to stay within
+                        their limits.
+                    - 'all': Shortcut for all states are observed to stay within the limits.
+
+            additional_constraints(list(Constraint)):
+                 Further constraints that shall be monitored. These have to be initialized first and passed to the
+                 ConstraintMonitor
+            merge_violations('max'/'product'/callable(*violation_degrees) -> float): Function to merge all single
+                violation degrees to a total violation degree.
+                    - 'max': Take the maximal violation degree as total violation degree.
+                    - 'product': Calculates the total violation degree as one minus the product of one minus all single
+                        violation degrees.
+                    - callable(*violation_degrees) -> float: User defined function to calculate the total violation.
+        """
+        self._constraints = list(additional_constraints)
+        if len(limit_constraints) > 0:
+            self._constraints.append(LimitConstraint(limit_constraints))
+
+        assert all(callable(constraint) for constraint in self._constraints)
+        assert merge_violations in ['max', 'product'] or callable(merge_violations)
+        self._merge_violations = lambda state: 0.0
+
+        if merge_violations == 'max':
+            self._merge_violations = max
+        elif merge_violations == 'product':
+            self._merge_violations = \
+                lambda violations: 1 - reduce(lambda val, violation: val * (1 - violation), violations, 0)
+        elif callable(merge_violations):
+            self._merge_violations = merge_violations

```

Comment:
assigning lambda funcs to variables is not encouraged practice.
Let's have real functions with "def" defined within this __init__ function